python/pyarrow/table.pxi

```diff
@@ -1513,12 +1513,18 @@ cdef class _Tabular(_PandasConvertible):
 
         return self.column(key)
 
+    def __len__(self):
+        return self.num_rows
+
     def __repr__(self):
         if not self._is_initialized():
             raise ValueError("This object's internal pointer is NULL, do not "
                              "use any methods or attributes on this object")
         return self.to_string(preview_cols=10)
 
+    def _column(self, int i):
+        raise NotImplementedError
+
```

Comment:
I plan to leave it as-is for now since it utilizes C objects (`self.table` or `self.batch`). I haven't found a way to consolidate C/Cython code since that code requires the explicit object types in order to compile. If we found a way, we could consolidate a large chunk of Cython! It might not be possible, or if it is possible it might give worse performance.

At this point, I think the best next steps are to leave the C/Cython as-is and focus on implementing missing APIs that might exist in one class and not the other.

Reply:
Thanks so much for the explanation! Helped me understand this part much better 👍